### ISSUE TITLE
so <- sleuth_fit(so, ~condition, 'full') error: Error in if (sum(valid) == 0) { : missing value where TRUE/FALSE needed

### ISSUE THREAD

**Issue by tiramisutes** [pachterlab/sleuth#125]
Hi,
I used this sleuth for 6 samples that no replicates as follows:
```
sample condition
HR-1B.fq.gz  HR1
HR-2B.fq.gz  HR2
HR-3B.fq.gz  HR3
HY-1B.fq.gz  HY1
HY-2B.fq.gz  HY2
HY-3B.fq.gz  HY3
```
This is my R code used and how to fit the error about `so <- sleuth_fit(so, ~condition, 'full')`? Thanks.
```
> so <- sleuth_fit(so, ~condition, 'full')
fitting measurement error models
shrinkage estimation
Error in if (sum(valid) == 0) { : missing value where TRUE/FALSE needed
> so <- sleuth_fit(so, ~1, 'reduced')
fitting measurement error models
shrinkage estimation
computing variance of betas
> so <- sleuth_lrt(so, 'reduced', 'full')
Error in model_exists(obj, alt_model) : model 'full' not found
> models(so)
[  reduced  ]
formula:  ~1 
coefficients:
        (Intercept)
```

**Comment by warrenmcg** [comment]
It would be helpful to us if you could help us pinpoint the cause of the error. There were no warnings during `sleuth_prep`?

After you have done `sleuth_prep`, could you do the following lines of code?

```
X <- model.matrix(formula("~condition"), so$sample_to_covariates)
rownames(X) <- so$sample_to_covariates$sample
A <- solve(t(X) %*% X)

mes <- sleuth:::me_model_by_row(so, X, so$bs_summary)
tid <- names(mes)
mes_df <- dplyr::bind_rows(lapply(mes,
    function(x) {
      data.frame(rss = x$rss, sigma_sq = x$sigma_sq, sigma_q_sq = x$sigma_q_sq,
        mean_obs = x$mean_obs, var_obs = x$var_obs)
    }))
mes_df$target_id <- tid
rm(tid)

mes_df <- dplyr::mutate(mes_df, sigma_sq_pmax = pmax(sigma_sq, 0))
data <- as.data.frame(mes_df)
data <- dplyr::mutate(data, x_ecdf = ecdf(data[, "mean_obs"])(data[, "mean_obs"]))
data <- dplyr::mutate(data, x_group = cut(x_ecdf, n_bins))
data <- dplyr::mutate(data, label = as.numeric(factor(x_group)))

res <- sapply(seq(length(levels(x_group))), function(x) {
  subset <- data[data$label == x, ]
  return(nrow(subset))
}

bad_x_group_lvls <- levels(x_group)[which(res == 0)]
print(bad_x_group_lvls)
```

It will also be helpful to us if you can send the `data` object. Do the following:

```
write.table(data, "test_data.txt", sep = "\t", row.names = F, quote = F)
```

Then compress the `test_data.txt` file (zip, gzip, tar/bzip, etc) and then post the resulting file following [these](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) instructions.

Thanks!

**Comment by tiramisutes** [comment]
Yes, there is no any warnings during  ``sleuth_prep`` and all is normal.  Actually, when I only select two samples and copy them like follows, it can work.  
```
sample condition
HR-1B.fq.gz HR1
HR-1B.fq.gz2 HR1
HR-2B.fq.gz HR2
HR-2B.fq.gz2 HR2
```

But, there are nearly 87% significant different expression genes (q-value <= 0.05) when I ``dim(significant genes)``. So, it's normal?
```
> so <- sleuth_prep(s2c, extra_bootstrap_summary = TRUE)
reading in kallisto results
dropping unused factor levels
....
normalizing est_counts
46861 targets passed the filter
normalizing tpm
merging in metadata
summarizing bootstraps
....

> dim(sleuth_significant)
[1] 41147    13
```
Besides, if I need remove the zero TPM in kallisto resulted file ``abundance.tsv``?
```
target_id	length	eff_length	est_counts	tpm
[HR-1B]_37350	1774	1595	3379.23	138.74
[HR-1B]_40861	1656	1477	0	0
[HR-1B]_49471	1426	1247	98.763	5.18646
```

**Comment by warrenmcg** [comment]
It's not surprising that you got so many hits in your second set of samples; the groups consist of two samples that are exact copies, so there is no variation at all within each group. Even a small amount of variation between the two samples will thus result in "significant changes".

I think I've realized the error: I didn't notice until now that you do not have replicates. You cannot do any fitting using `condition` as a covariate because there is no way to estimate within-group variance with only one replicate per group. The `~1` model works because you're treating the entire cohort of six samples as one group, and you are only measuring within-group variance.

If you run my code above (modified from before), you'll notice that the residual sum of squares (RSS), and the variance (sigma_sq) are 0 and `NA`, respectively.

It seems like the first three samples and the second three samples are related; is that true? Could you rename the condition column using just the "HR" and the "HY" labels? You will get meaningful results from that, but not from your current sample_to_covariates table.

**Comment by warrenmcg** [comment]
pinging @pimentel: should there be an informative error thrown when people attempt to fit a model which has no replicates in a group? Am I interpreting this situation correctly?

**Comment by pimentel** [comment]
@warrenmcg indeed, the issue is that there is are no replicates. also, you are correct that we should definitely output a better message.

**Comment by tiramisutes** [comment]
Now, I understand what you explain and know my error output. Yes, the first three samples and the second three samples are related, but all these six samples are different and **I want to find the significant differentially expressed genes between the first three samples and the second three samples, and the not significantly differentially expressed genes within the first three samples or the second three samples.**  How to do it?

**Comment by warrenmcg** [comment]
Well, if you relabel the condition column by dropping the numbers:
```
sample condition
HR-1B.fq.gz  HR
HR-2B.fq.gz  HR
HR-3B.fq.gz  HR
HY-1B.fq.gz  HY
HY-2B.fq.gz  HY
HY-3B.fq.gz  HY
```

The statistics "under the hood" of sleuth looks at the overall variance of a gene, and then first attempts to estimate what part of that variance is from the "technical variance" introduced by kallisto, and then compares the variance within a group to the variance between groups. The latter is what the `wald` test looks at. The `lrt` (likelihood ratio test) compares two separate but related linear models, and tries to see whether the more complicated model explains the variance better than the simpler model. If the LRT test is significant, one makes the decision that the more complicated model is capturing the variance better than the simpler model. 

So, in conclusion, using my suggested modified `s2c` table and the normal pipeline (in particular, using the default `wald` test because your experimental design has just one covariate) should address your question.

Hope that helps!

**Comment by warrenmcg** [comment]
@pimentel, my testing indicates that no error is thrown when at least one factor level has two samples. Do we want to throw an error only when all factor levels have no replicates, or do we want to also throw a warning when any factor level has just 1 replicate?

**Comment by tiramisutes** [comment]
Yes, I think I also can get the not significantly differentially expressed genes within the first three samples through first copy them to calculate ``sleuth_significant`` and then opposite side of this resulted.

**Comment by vasanthmandla** [comment]
Hi I used the below samples to perform sleuth for deferentially expressed genes to between two conditions (Control and Treatment) with 4 replicates each.

 sample  condition
1  Ctrl1   Control1
2  Ctrl2   Control2
3  Ctrl3   Control3
4  Ctrl4   Control4
5   FAC_1 Treatment1
6   FAC_2 Treatment2
7   FAC_3 Treatment3
8   FAC_4 Treatment4

I am having the same error  
Error in if (sum(valid) == 0) { : missing value where TRUE/FALSE needed
when I ran the R code
 so <- sleuth_fit(so, ~condition, 'full')

**Comment by warrenmcg** [comment]
@vasanthmandla When you look at the condition column, you'll notice that each sample has a unique condition associated with it, which yields an invalid model for the sleuth linear modeling. Based on the discussion above, you should change your condition column to read

```
sample condition
1 Ctrl1 Control
2 Ctrl2 Control
3 Ctrl3 Control
4 Ctrl4 Control
5 FAC_1 Treatment
6 FAC_2 Treatment
7 FAC_3 Treatment
8 FAC_4 Treatment
```

when you do the testing, the beta would be "conditionTreatment". Hope that helps. This issue will remain closed since the issue has already been solved.

**Comment by vasanthmandla** [comment]
@warrenmcg Thank you very much for the clarification. It did work but not I am another error. I tried to find the cause but I couldn't. Kindly help me to resolve it.

so <- sleuth_lrt(so, 'reduced', 'full')
#Error in model_exists(obj, alt_model) : model 'full' not found
#Calls: sleuth_lrt -> model_exists
#Execution halted

I am sorry if I am not supposed to post the errors over here. Thanks.

**Comment by bioinfo-dirty-jobs** [comment]
I have the same error:

```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.5 LTS

Matrix products: default
BLAS: /home/maurizio/miniconda2/envs/kalisto/lib/R/lib/libRblas.so
LAPACK: /home/maurizio/miniconda2/envs/kalisto/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=it_IT.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=it_IT.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=it_IT.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=it_IT.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] bindrcpp_0.2.2 sleuth_0.30.0 

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0           BiocManager_1.30.4   pillar_1.3.0        
 [4] compiler_3.5.1       plyr_1.8.4           bindr_0.1.1         
 [7] tools_3.5.1          prettyunits_1.0.2    progress_1.2.0      
[10] bitops_1.0-6         biomaRt_2.36.1       digest_0.6.18       
[13] bit_1.1-12           RSQLite_2.1.1        memoise_1.1.0       
[16] tibble_1.4.2         gtable_0.2.0         rhdf5_2.26.0        
[19] pkgconfig_2.0.2      rlang_0.3.0.1        DBI_1.0.0           
[22] curl_3.2             parallel_3.5.1       httr_1.3.1          
[25] stringr_1.3.1        dplyr_0.7.8          hms_0.4.2           
[28] IRanges_2.14.12      S4Vectors_0.18.3     stats4_3.5.1        
[31] bit64_0.9-7          grid_3.5.1           tidyselect_0.2.5    
[34] Biobase_2.40.0       glue_1.3.0           data.table_1.11.6   
[37] R6_2.2.2             AnnotationDbi_1.42.1 tcltk_3.5.1         
[40] XML_3.98-1.16        ggplot2_3.1.0        Rhdf5lib_1.4.2      
[43] purrr_0.2.5          blob_1.1.1           magrittr_1.5        
[46] BiocGenerics_0.26.0  scales_1.0.0         assertthat_0.2.0    
[49] colorspace_1.3-2     stringi_1.2.4        RCurl_1.95-4.11     
[52] lazyeval_0.2.1       munsell_0.5.0        crayon_1.3.4  
```

**Comment by nhuhoa** [comment]
I also got the same issue, and after debugging program, I know where the issue is. I run sleuth with only 2 samples without replicates (as authors pointed out above), each sample is one condition, so the design formula X is: 

sample condition
SA53   Condition1
SA54   Condition2

X <- model.matrix(formula, obj$sample_to_covariates)
# X will have 2 rows and 2 columns

In the sleuth 'me_model' function:
n <- nrow(X)  # value is 2
degrees_free <- n - ncol(X) # value is 0 because ncol(X) is also equal 2
ols_fit <- lm.fit(X, y)
rss <- sum(ols_fit$residuals^2)
sigma_sq <- rss/(degrees_free) - sigma_q_sq  #degrees_free is 0 so this divide by zero --> NA values for all sigma_sq 
Following this, function me_model_by_row returns all NA values for sigma_sq, NA values for sigma_sq_pmax, and then error "Error in if (sum(valid) == 0)" at sliding_window_grouping()

I hope authors can help us to fix error of divide by zeros above. 

Thanks
